### ISSUE TITLE
Focus is lagging when it's minimized for a long time

### ISSUE THREAD

**Issue by Roman-Skabin** [focus-editor/focus#94]
**Describe the bug**
When Focus is minimized for about 30-60 minutes. It starts lagging very heavily.

**To Reproduce**
1. Start Focus
2. Open a project
3. Minimize window
4. Hold it minimized for about 30-60 minutes
5. Try to type something or search for something, do whatever you want with a text.

**Expected behavior**
Focus is not lagging

- OS: Windows 10 Enterprise 22H2, Windows 11 Pro latest
- Focus Version: 0.1.10

**Comment by focus-editor** [comment]
I've tried following the provided steps but couldn't reproduce the behavior. I've also tried to change the files externally while the editor is minimized, to no avail.

Are you in the beta? It would be really helpful if you could use a profiler on a debug version of the editor to see exactly where the time is spent when it's lagging.
Alternatively I could give you a debug executable together with a pdb file.

**Comment by Roman-Skabin** [comment]
Yeah, I'm in the beta. You want me to use a memory profiler from Basic or your custom one?

**Comment by focus-editor** [comment]
I was thinking more of a general sampling profiler like Very Sleepy - you could just run the editor with it or attach to the process and then profile for a few seconds when it starts lagging. That should give us an idea what code is causing the lagging.

Please profile all threads. Thanks

**Comment by Flamaros** [comment]
I think I got a pretty similar issue today. I let focus open on my office computer during a day off, when coming back to the office this morning I saw focus unresponsive to Windows events (window moving, resizing, closing,...) and completely empty (I had only the Windows tile bar and the background color (dark blue).

Process Explorer was indicating that focus was using 11GB of RAM, so even if I have 32 GB on this computer it is pretty impressive. I have open only few small text files, so I think it can be only due to the memory leak that focus become unresponsive.

I have the latest version of focus (v0.2.0).

**Comment by focus-editor** [comment]
@Flamaros  The increased RAM usage suggests that you've hit a bug where it scanned the whole directory containing one of the files you've opened. If that happens again you can verify that by pressing Ctrl-P and checking what files are open - it's an elusive bug which still happens sometimes.

This is unlikely to be related to minimizing.

**Comment by Flamaros** [comment]
> @Flamaros The increased RAM usage suggests that you've hit a bug where it scanned the whole directory containing one of the files you've opened. If that happens again you can verify that by pressing Ctrl-P and checking what files are open - it's an elusive bug which still happens sometimes.
> 
> This is unlikely to be related to minimizing.

Ok, I created a new issue for memory usage: https://github.com/focus-editor/focus/issues/184

After a fresh start even if focus take 12GB of memory it stay pretty responsive. You are right the unresponsiveness issue is different and is certainly just related to the running time of focus.

**Comment by Roman-Skabin** [comment]
Focus was opened for a long time and now it's lagging during scrolling.
Very Sleepy capture: [capture.zip](https://github.com/focus-editor/focus/files/13490068/capture.zip)

\------------------------------
Version: commit 6019a9f9cb5e840f2196c769cbdb25871796a775

**Comment by Roman-Skabin** [comment]
I will attach main thread capture if I reproduce it once again

**Comment by Roman-Skabin** [comment]
Here is main thread capture: [main_thread.zip](https://github.com/focus-editor/focus/files/13528074/main_thread.zip)

\------------------------------
Version: f4d77351dd90c5b66913836811e9fe7e42b0a793 commit

**Comment by focus-editor** [comment]
Thanks, that could be it. Will try without it.

**Comment by Roman-Skabin** [comment]
I meant to remove all these 3 lines completely from the if-case. You do not need to call ValidateRect unless you use GDI.

**Comment by focus-editor** [comment]
If we remove those 3 lines completely, then WM_PAINT will be handled by the default case.
Also, if you don't call ValidateRect, nothing will be drawn and the program will hang. I don't know Win32 well enough to understand why. 
If you know why please let me know

**Comment by Roman-Skabin** [comment]
The issue is not fixed 😞. I've profiled the editor while it was lagging via visual studio profiler and it says the following:
![image](https://github.com/focus-editor/focus/assets/40520199/059d6a89-4935-4a4b-af0e-475f9e1da098)
It seems like we spend a lot of time in some call inside immediate_set_shader.

VerySleepy doesn't say much as well as any other sampling profiler. Do you have any tools inside focus itself to detect the problem? All I've found is DUMP_GL_ERRORS in GL module, but as far as I can see it will not help much.

**Comment by Roman-Skabin** [comment]
Btw, VerySleepy also says that we spend a lot of time in nvoglv64 DrvValidateVersion (which is not focus problem, obviously). All I found is this discussion, but it's not very useful. https://forums.developer.nvidia.com/t/mysterious-cpu-usage-issue/53710

**Comment by focus-editor** [comment]
Let's fix the obvious problems with `immediate_set_shader` first and see if that helps.

I'm thinking of moving to DirectX and get rid of Simp as it currently is. OpenGL drivers seem to be more flaky

**Comment by Roman-Skabin** [comment]
Yesterday I had Focus being opened for a whole day on background and it wasn't lagging anymore. It seems like either it's fixed or either the lag is reduced/minimized

**Comment by Roman-Skabin** [comment]
Here we go again
![image](https://github.com/focus-editor/focus/assets/40520199/e7d55359-0db4-4733-90bd-31d18956b14e)

set_parameter:
![image](https://github.com/focus-editor/focus/assets/40520199/4ed0c589-8e94-4e8b-889c-4baa460321b0)

It seems like it's an OpenGL driver who kills my system

**Comment by focus-editor** [comment]
Will fix this call in the same manner, thanks!

**Comment by focus-editor** [comment]
Unfortunately I wasn't yet able to fix it because OpenGL is not cooperating

**Comment by focus-editor** [comment]
Setting explicit uniform locations is only supported since OpenGL 4.3 as it seems, and we can't really use it because macOS only supports up to 4.1, so will have to find another solution.

This is another reason to migrate away from OpenGL

**Comment by Roman-Skabin** [comment]
You can get their location only once on start and save them for future use

**Comment by focus-editor** [comment]
@Roman-Skabin sorry it's been neglected for so long, I somehow managed to forget about it. Hopefully this should fix the issues you were seeing.

**Comment by Flamaros** [comment]
I still have the issue (tried with version 0.3.5).

I put a very sleepy trace.
I will eventually try to compare the trace with one of my own programs, because it seems there is something weird with how the nvidia driver reacts to the swapbuffer.

Sadly I don't have access to the beta of jai, so I can't debug it myself.


[focus-editor.zip](https://github.com/focus-editor/focus/files/15345264/focus-editor.zip)

**Comment by ileonte** [comment]
This **might** be fixed in the nightlies (more specifically by this commit: https://github.com/focus-editor/focus/commit/0777040f33eea2c02428791c9cd9beb665da8a20) - there was an indirect DC leak caused by Focus issuing draw commands while the window was minimized.

Please try the latest nightly build if you can.

**Comment by Roman-Skabin** [comment]
Actually, I haven't experienced that behaviour for a long time already, so I'm not sure I will notice any difference.

**Comment by Flamaros** [comment]
Oh thx, I'll see if I can try.

**Comment by Flamaros** [comment]
After a day, it is still good. I let him run in background and I'll check in few days.

**Comment by Flamaros** [comment]
I took a look at it today, and IMO the issue is fixed. So I'll consider to migrate from VS to focus for my day job too.

I let you close the issue.

**Comment by focus-editor** [comment]
Cool!
